theories/crypto/LorR.ec

```diff
@@ -29,46 +31,46 @@ section.
 
   local module Aux = {
     var b,r : bool
-    proc main () = { 
+    proc main (x:input) = { 
       b <${0,1};
-      if (b)  r <@ L.main();
-      else r <@ R.main ();
+      if (b)  r <@ L.main(x);
+      else r <@ R.main (x);
       return (b = r);
     }
   }.
 
-  lemma pr_AdvLR_AdvRndLR &m: 
-     Pr[R.main() @ &m : true] = 1%r =>
-     `| Pr[L.main () @ &m : res] - Pr[R.main () @ &m : res] | =
-     2%r * `| Pr[RandomLR(L, R).main() @ &m : res] - 1%r/2%r |.
+  lemma pr_AdvLR_AdvRndLR &m (x' : input): 
+     Pr[R.main(x') @ &m : true] = 1%r =>
+     `| Pr[L.main (x') @ &m : res] - Pr[R.main (x') @ &m : res] | =
+     2%r * `| Pr[RandomLR(L, R).main(x') @ &m : res] - 1%r/2%r |.
   proof.
     move => Hll.
-    have -> : Pr[RandomLR(L, R).main() @ &m : res] = Pr[Aux.main() @ &m : res].
+    have -> : Pr[RandomLR(L, R).main(x') @ &m : res] = Pr[Aux.main(x') @ &m : res].
     + by byequiv => //; proc; sim=> /#.
-    have -> : Pr[Aux.main() @ &m : res] = 
-              Pr[Aux.main() @ &m : res /\ Aux.b] +  
-              Pr[Aux.main() @ &m : res /\ !Aux.b].
+    have -> : Pr[Aux.main(x') @ &m : res] = 
+              Pr[Aux.main(x') @ &m : res /\ Aux.b] +  
+              Pr[Aux.main(x') @ &m : res /\ !Aux.b].
     + by rewrite Pr [mu_split Aux.b].
-    have -> : Pr[Aux.main() @ &m : res /\ Aux.b] = 1%r/2%r * Pr[L.main () @ &m : res].
-    + byphoare (_: (glob L) = (glob L){m} ==> res /\ Aux.b) => //.
-      proc; pose p := Pr[L.main() @ &m : res]. (* FIXME assert false without the pose *)
-      seq 1 : (Aux.b = true) (1%r/2%r) p  _ 0%r (glob L = (glob L){m}); 1:by auto.
+    have -> : Pr[Aux.main(x') @ &m : res /\ Aux.b] = 1%r/2%r * Pr[L.main (x') @ &m : res].
+    + byphoare (_: (glob L) = (glob L){m} /\ x = x' ==> res /\ Aux.b) => //.
+      proc; pose p := Pr[L.main(x') @ &m : res]. (* FIXME assert false without the pose *)
+      seq 1 : (Aux.b = true /\ x = x') (1%r/2%r) p  _ 0%r (glob L = (glob L){m} /\ x = x'); 1:by auto.
       + by rnd (pred1 true); skip => />; rewrite dbool1E.
       + if; last by (conseq (_: false ==> _); 1:by smt()); auto.
         conseq (_ : _ ==> Aux.r) => //; 1:by smt().
-        call (_: glob L = (glob L){m} ==> res); last by auto.
-        by bypr=> &m0 hm0; rewrite /p; byequiv=> //; proc (true).
+        call (_: glob L = (glob L){m} /\ x = x' ==> res). 
+        bypr=> &m0 hm0; rewrite /p; byequiv=> //; proc (true);smt(). trivial.  
```

Comment:
This isn't nice. Doest the `last by auto` not solve it?
Do we really need to call out to `smt` for something that looks like a sequence of substitutions?
Please put the `by` back at the front of the line if it closes a goal.

Reply:
Ah! The `last` was optional, it could have been some other focus selector... No matter.